cobc/replace.c

```diff
@@ -0,0 +1,687 @@
+			const struct cb_replace_list *replace_list)
+{
+#ifdef DEBUG_REPLACE_TRACE
+	fprintf(stderr, "%scheck_replace_all(%s,",
+		DEPTH, repls->name);
+	fprintf(stderr, "%s    new_text = %s,\n", DEPTH,
+		string_of_list(new_text));
+	fprintf(stderr, "%s    tokens = %s,\n", DEPTH,
+		string_of_list(tokens));
+	fprintf(stderr, "%s    src = %s,\n", DEPTH,
+		string_of_list(src));
+	fprintf(stderr, "%s)\n", DEPTH);
+#endif
+
+	if (src==NULL){
+		/* MATCH */
+		ppecho_switch_list (MORE_DEPTH repls, new_text) ;
+		repls->token_queue = tokens ;
+		check_replace_after_match (MORE_DEPTH repls);
+	} else {
+		const char* src_token = src->text;
+		if (src_token[0] == ' ' || src_token[0] == '\n'){
+			check_replace_all (MORE_DEPTH repls,new_text,tokens,
+					   src->next, replace_list);
+		} else {
+			if (tokens == NULL){
+				/* PARTIAL MATCH, wait for next token */
+#ifdef DEBUG_REPLACE_TRACE
+	fprintf(stderr, "%s  check_replace_all --> PARTIAL MATCH\n", DEPTH);
+#endif
+			} else {
+				const char* token = tokens->text;
+				tokens = tokens->next;
+				if (token[0] == ' ' || token[0] == '\n'){
+					check_replace_all (MORE_DEPTH repls,
+							   new_text,
+							   tokens, src,
+							   replace_list);
+				} else {
+					if (!strcasecmp(src_token,token)){
+
+						check_replace_all(
+							MORE_DEPTH repls,
+							new_text,
+							tokens,src->next,
+							replace_list);
+					} else {
+						check_replace (
+							MORE_DEPTH repls,
+							replace_list);
+					}
+				}
+			}
+		}
+	}
+}
+
+static
+void check_replace_after_match (WITH_DEPTH struct cb_replacement_state *repls)
+{
+#ifdef DEBUG_REPLACE_TRACE
+	fprintf(stderr, "%scheck_replace_after_match(%s)\n",
+		DEPTH, repls->name);
+#endif
+  repls->current_list = NULL;
+  if (repls->token_queue != NULL){
+	  if( repls->token_queue->text[0] == ' ' || 
+	      repls->token_queue->text[0] == '\n' ){
+		  ppecho_switch (MORE_DEPTH repls,
+				 repls->token_queue->text);
+		  repls->token_queue = repls->token_queue->next;
+		  check_replace_after_match (MORE_DEPTH repls);
+	  } else {
+		  do_replace (MORE_DEPTH repls);
+	  }
+  }
+}
+
+static
+void do_replace (WITH_DEPTH struct cb_replacement_state* repls)
+{
+#ifdef DEBUG_REPLACE_TRACE
+	fprintf(stderr, "%sdo_replace(%s)\n",DEPTH, repls->name);
+#endif
+	if (repls->current_list == NULL){
+		if (repls->replace_list == NULL){
+
+			/* Beware: this is incorrect if a REPLACE is
+			 * withing the queue, as it has already been
+			 * parsed before any COPY-REPLACING
+			 * substitution. */
+			ppecho_switch_list (MORE_DEPTH repls,
+					    repls->token_queue);
+			repls->token_queue = NULL;
+		} else {
+			check_replace (MORE_DEPTH repls, repls->replace_list);
+		}
+	} else {
+		check_replace (MORE_DEPTH repls, repls->current_list);
+	}
+}
+
+/* Whether a word matches the definition of WORD in pplex.l */
+static
+int is_word (WITH_DEPTH const char* s){
+	int i;
+	int len = strlen (s);
+
+
+	for( i = 0; i<len ; i++){
+		char c = s[i];
+		if( c == '_'
```

Comment:
Shouldn't this just check for `c != ',' &&  c != ';' && !isspace(c)`?

Reply:
This function is used to merge multiple WORD tokens into a single WORD token. So, I followed the definition of WORD here: https://github.com/OCamlPro/gnucobol/blob/516b813b3ca05ecbc732efdf359ca77db89af449/cobc/pplex.l#L191